conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/Parameters.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.lib;
+
+import com.google.common.collect.Multimap;
+import com.palantir.logsafe.SafeArg;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.AttachmentKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class to put different types of safe or unsafe parameters in a MultiMap attached to the exchange.
+ * It requires the exchange to have been priorly through a ParametersHandler which would have created the attached
+ * MultiMaps.
+ */
+public final class Parameters {
+
+    private static final Logger log = LoggerFactory.getLogger(Parameters.class);
+
+    public static final AttachmentKey<Multimap<String, Object>> SAFE_PARAMS_ATTACH_KEY =
```

Comment:
Do these need to be multimaps? If we key by the variable name we're guaranteed uniqueness because a method may not have multiple parameters with the same name.

Reply:
There is only one case where we are not guaranteed, headers have the name of the actual header name and not the header argument. I am also potentially interested to stash more things into the params thing (see the WC PR #1303)

Reply:
bikeshed: The more I think about this, I'd prefer a standard map because we record this data in a standard map. My preference is to throw an exception when there are multiple attempts to record data under the same name.

This allows us to solve the multi-value problem where currently WC logs `["foo"]` as `"foo"` because it assumes everything might be a list. We've generated the code to solve this problem already, and we should record exactly the received value, avoiding datatype mismatches depending on the number of parameters provided.

Headers: We can use the header parameter name rather than the header name.

Alternative solution:

We can entirely avoid logic here, and change the attachment to `List<Arg<?>>` which does allow duplicate keys, and allow the consumer (request logging implementation) to decide how to handle that data.

Reply:
By throwing exception, you mean at conjure generator time, or at runtime? Would that mean the endpoint always error out?

re: headers, if we use the header parameter name then we cannot use the whitelist anymore as a safe guard but maybe we don't need that additional protection.

The `List<Arg>` is interesting, maybe that's actually the best solution after all.